[Bundle 2/n] Prune g_chainman usage in mempool-related validation functions

dongcarl opened the pull request:
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

dongcarl committed:
validation: Pass in chainstate to AcceptToMemoryPoolWithTime

dongcarl committed:
validation: Pass in chainstate to ::LoadMempool

dongcarl committed:
validation: Pass in coins cache to ::LimitMempoolSize

dongcarl committed:
validation: Pass in chainstate to IsCurrentForFeeEstimation

dongcarl committed:
validation: Pass in coins tip to CheckInputsFromMempoolAndCache

dongcarl committed:
validation: Add chainstate member to MemPoolAccept

dongcarl committed:
validation: Pass in chainstate to ::AcceptToMemoryPool

dongcarl committed:
scripted-diff: Invoke ::AcceptToMemoryPool with chainstate

-BEGIN VERIFY SCRIPT-
find_regex='\bAcceptToMemoryPool\(' \
    && git grep -l -E "$find_regex" -- src \
        | grep -v '^src/validation\.\(cpp\|h\)$' \
        | xargs sed -i -E 's@'"$find_regex"'@\0::ChainstateActive(), @g'
-END VERIFY SCRIPT-

dongcarl committed:
validation: Pass in chain tip to ::CheckFinalTx

dongcarl committed:
scripted-diff: Invoke ::CheckFinalTx with chain tip

-BEGIN VERIFY SCRIPT-
find_regex='\bCheckFinalTx\(' \
    && git grep -l -E "$find_regex" -- src \
        | grep -v '^src/validation\.\(cpp\|h\)$' \
        | xargs sed -i -E 's@'"$find_regex"'@\0::ChainActive().Tip(), @g'
-END VERIFY SCRIPT-

dongcarl committed:
validation: Pass in chainstate to ::CheckSequenceLocks

dongcarl committed:
validation: Pass in chain to ::TestLockPointValidity

dongcarl committed:
validation: Pass in chainstate to CTxMemPool::removeForReorg

Several other parameters are now redundant since they can be safely
obtained from the chainstate given that ::cs_main is locked. These are
now removed.

dongcarl committed:
validation: Pass in chainstate to UpdateMempoolForReorg

dongcarl committed:
validation: Use *this in CChainState::LoadMempool

dongcarl committed:
tree-wide: Fix erroneous AcceptToMemoryPool replacements

dongcarl committed:
validation: Remove old CheckFinalTx w/o chain tip param

dongcarl committed:
validation: Remove old AcceptToMemoryPool w/o chainstate param

jnewbery commented:
https://github.com/bitcoin/bitcoin/pull/20749 is merged. Rebase please :slightly_smiling_face:

dongcarl commented:
Pushed 5803d30babf19a86828b1d967b0d38ee8c6551c4 -> 8d435f277f44dbd6807c809c34411b6f1f85b4bc
- Rebased over master since "#20749 | [Bundle 1/n] Prune g_chainman usage related to ::LookupBlockIndex" is now merged!

ariard commented on src/test/validation_block_tests.cpp:
```diff
@@ -286,7 +286,7 @@ BOOST_AUTO_TEST_CASE(mempool_locks_reorg)
             std::list<CTransactionRef> plTxnReplaced;
             for (const auto& tx : txs) {
-                BOOST_REQUIRE(AcceptToMemoryPool(
+                BOOST_REQUIRE(AcceptToMemoryPool(::ChainstateActive(), 
```
whitespace?

dongcarl replied:
This is fixed in the subsequent commit!

ariard commented on src/txmempool.cpp:
```diff
@@ -356,7 +356,7 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
 {
     // Add to memory pool without checking anything.
-    // Used by AcceptToMemoryPool(), which DOES do
+    // Used by AcceptToMemoryPool(::ChainstateActive(), ), which DOES do
```
I think the scripted-diff is too much expansive here.

EDIT: note for reviewers, fixed in next commit.

dongcarl replied:
Yeah I should have used a clang-tidy script for this... Oh well, perhaps we can set that up in the future.

ariard reviewed (commented):
Approach ACK, did a first parse on each commit.

Side-question, what's your heuristic to divide your bundles ?

dongcarl commented:
Pushed 8d435f277f44dbd6807c809c34411b6f1f85b4bc -> 64db9d61f4c23acb8b68b1a778db02aed0f582f1
- Rebased over master for #21051
- Addressed https://github.com/bitcoin/bitcoin/pull/20750#discussion_r568151119

dongcarl commented on src/validation.cpp:
```diff
@@ -202,9 +202,10 @@ static FILE* OpenUndoFile(const FlatFilePos &pos, bool fReadOnly = false);
 static FlatFileSeq UndoFileSeq();
 
-bool CheckFinalTx(const CTransaction &tx, int flags)
+bool CheckFinalTx(CBlockIndex* active_chain_tip, const CTransaction &tx, int flags)
```
Done! See: e5198eff4b80d0d39d1cd9b3da317b25026c436d

glozow reviewed (commented):
Concept ACK, planning to review more closely soon. 
~Would it ever be the case that we call `ATMP(chainstate, pool...)` where `pool` != `chainstate.m_mempool`? If not, it seems like you could take `pool` out of the function signature(s) as well.~ (oh jk 😳 it's protected)

dongcarl commented:
> Side-question, what's your heuristic to divide your bundles ?

It's mostly bundles of functions which call each other, not too scientific, but works!

ryanofsky commented on src/validation.cpp:
```diff
@@ -458,11 +458,13 @@ namespace {
         m_limit_descendants(gArgs.GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT)),
-        m_limit_descendant_size(gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000) {}
+        m_limit_descendant_size(gArgs.GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000) {
+        assert(std::addressof(::ChainstateActive()) == std::addressof(m_active_chainstate));
```
In commit "validation: Add chainstate member to MemPoolAccept" (dd7e989b0a4ad6949ccb70d13405f7af3f05b724)

Assert might be more meaningful if it were added all the places the new m_active_chainstate member is used

dongcarl replied:
Agreed! Done in: 76a4aa60dbc8e142a72c84cf35ebc04c7654866f

ryanofsky reviewed (approved):
Code review ACK 64db9d61f4c23acb8b68b1a778db02aed0f582f1

jnewbery commented on src/validation.h:
```diff
@@ -228,7 +228,7 @@ bool TestLockPointValidity(const LockPoints* lp) EXCLUSIVE_LOCKS_REQUIRED(cs_mai
  * See consensus/consensus.h for flag definitions.
  */
-bool CheckSequenceLocks(const CTxMemPool& pool, const CTransaction& tx, int flags, LockPoints* lp = nullptr, bool useExistingLockPoints = false) EXCLUSIVE_LOCKS_REQUIRED(::cs_main, pool.cs);
+bool CheckSequenceLocks(CChainState& active_chainstate, const CTxMemPool& pool, const CTransaction& tx, int flags, LockPoints* lp = nullptr, bool useExistingLockPoints = false) EXCLUSIVE_LOCKS_REQUIRED(::cs_main, pool.cs);
```
I find these 200+ character lines very difficult to parse. It'd be nice if PRs didn't make them even longer and more unreadable. Consider something like:

```suggestion
bool CheckSequenceLocks(CChainState& active_chainstate, const CTxMemPool& pool,
                        const CTransaction& tx, int flags, LockPoints* lp = nullptr,
                        bool useExistingLockPoints = false)
    EXCLUSIVE_LOCKS_REQUIRED(::cs_main, pool.cs);
```

(same for the function definition in validation.cpp)

dongcarl replied:
Done in: bf2927643162c650f2a54dd33a83eccaa67ac50c

jnewbery commented on src/validation.h:
```diff
@@ -184,7 +184,7 @@ void PruneBlockFilesManual(int nManualPruneHeight);
  * plTxnReplaced will be appended to with all transactions replaced from mempool
  * @param[out] fee_out optional argument to return tx fee to the caller **/
-bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
+bool AcceptToMemoryPool(CChainState& active_chainstate, CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
```
Again, this change makes it harder to read this function declaration.

dongcarl replied:
Done in: 2eceb88e52e06e5d58c864b67ff65bc91749eeb1, added a not to myself to review the future commits for this!

jnewbery commented on src/txmempool.cpp:
```diff
@@ -499,16 +499,17 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
 }
 
-void CTxMemPool::removeForReorg(const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags)
+void CTxMemPool::removeForReorg(CChainState& active_chainstate, const CCoinsViewCache *pcoins, unsigned int nMemPoolHeight, int flags)
```
There's no need to pass in a `CCoinsViewCache *` or `nMemPoolHeight` here any more. You can just get them from the `CChainState&`.

dongcarl replied:
Done in: 6dd22bc4b052f42cb9d102e940a4196d62c975fb
Much cleaner!

jnewbery reviewed (commented):
Code review ACK 64db9d61f4c23acb8b68b1a778db02aed0f582f1

> Would it ever be the case that we call ATMP(chainstate, pool...) where pool != chainstate.m_mempool? If not, it seems like you could take pool out of the function signature(s) as well.

I agree with this. Anywhere that you're passing a `CChainState&`, you can avoid also passing a `CTxMemPool&` (after making `CChainState.m_mempool` public). It wouldn't make sense to call one of these functions with a different mempool from the one associated with the chainstate (and besides, we only ever have one mempool).

glozow commented:
I have a question - when is it necessary to pass the entire `CChainState`? For most of these functions, it makes perfect sense to me, but in others, it seems like the function could just use a specific member of chainstate because their responsibility is narrower.

For example, in e83c8538ec022fa29bed2894471cdee8a974f66f, `CheckInputsFromMempoolAndCache()` just looks at the coins in `view` and checks that they're consistent with mempool and current `CoinsTip()`, then calls `CheckInputScripts()` (which should be stateless). So it really only needs an extra `CCoinsViewCache&` from `chainstate.CoinsTip()`.

In 76e0811629cfe94a19a2ee52aff51bcba073b55c, it seems to me that `CheckSequenceLocks()` only really needs the `CBlockIndex*` chain tip, similarly to `CheckFinalTx()`?

ariard commented:
Code Review ACK 64db9d6

dongcarl committed:
style-only: Make CheckSequenceLock signature readable

dongcarl committed:
style-only: Make AcceptToMemoryPool signature readable

dongcarl commented:
@glozow 

> I have a question - when is it necessary to pass the entire `CChainState`? For most of these functions, it makes perfect sense to me (and I think args should be removed when they can be retrieved from the chainstate), but in others, it seems like the function could just use a specific member of chainstate because their responsibility is narrower.

You're right! I think most of this is vestigial stuff from when I was trying to keep things safe and keep the code as close to existing code as possible. However, with enough asserts, we can be more confident that these changes are a-okay.

> For example, in [e83c853](https://github.com/bitcoin/bitcoin/commit/e83c8538ec022fa29bed2894471cdee8a974f66f), `CheckInputsFromMempoolAndCache()` just looks at the coins in `view` and checks that they're consistent with mempool and current `CoinsTip()`, then calls `CheckInputScripts()` (which should be stateless). So it really only needs an extra `CCoinsViewCache&` from `chainstate.CoinsTip()`.

Good catch! Fixed in: f6de0ae51e5b3b824e7eef6b988904b67cd3dab0

> In [76e0811](https://github.com/bitcoin/bitcoin/commit/76e0811629cfe94a19a2ee52aff51bcba073b55c), it seems to me that `CheckSequenceLocks()` only really needs the `CBlockIndex*` chain tip, similarly to `CheckFinalTx()`?

I think `CheckSequenceLocks()` needs both `active_chainstate.m_chain.Tip()` and `active_chainstate.CoinsTip()`? I may be missing something here.

glozow commented on src/validation.cpp:
```diff
@@ -327,7 +335,7 @@ bool CheckSequenceLocks(const CTxMemPool& pool, const CTransaction& tx, int flag
 static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consensus::Params& chainparams);
 
-static void LimitMempoolSize(CTxMemPool& pool, size_t limit, std::chrono::seconds age)
+static void LimitMempoolSize(CTxMemPool& pool, CChainState& chainstate, size_t limit, std::chrono::seconds age)
```
I thiiiiink this only needs the `CCoinsViewCache` CoinsTip?

glozow commented on src/validation.cpp:
```diff
@@ -366,10 +376,11 @@ static bool IsCurrentForFeeEstimation() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
  */
 
-static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransactions& disconnectpool, bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
+static void UpdateMempoolForReorg(CChainState& active_chainstate, CTxMemPool& mempool, DisconnectedBlockTransactions& disconnectpool, bool fAddToMempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main, mempool.cs)
```
same, `mempool` not needed if you can get it from `active_chainstate`

glozow commented on src/validation.cpp:
```diff
@@ -4991,7 +5025,7 @@ int VersionBitsTipStateSinceHeight(const Consensus::Params& params, Consensus::D
 static const uint64_t MEMPOOL_DUMP_VERSION = 1;
 
-bool LoadMempool(CTxMemPool& pool)
+bool LoadMempool(CTxMemPool& pool, CChainState& active_chainstate)
```
same, `pool` not needed if you can get it from `active_chainstate`

glozow commented on src/validation.h:
```diff
@@ -228,7 +233,12 @@ bool TestLockPointValidity(const LockPoints* lp) EXCLUSIVE_LOCKS_REQUIRED(cs_mai
  * See consensus/consensus.h for flag definitions.
  */
-bool CheckSequenceLocks(const CTxMemPool& pool, const CTransaction& tx, int flags, LockPoints* lp = nullptr, bool useExistingLockPoints = false) EXCLUSIVE_LOCKS_REQUIRED(::cs_main, pool.cs);
+bool CheckSequenceLocks(CChainState& active_chainstate,
```
same, `pool` not needed if you can get it from `active_chainstate`

glozow commented on src/validation.h:
```diff
@@ -184,9 +184,14 @@ void PruneBlockFilesManual(int nManualPruneHeight);
  * plTxnReplaced will be appended to with all transactions replaced from mempool
  * @param[out] fee_out optional argument to return tx fee to the caller **/
-bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
+bool AcceptToMemoryPool(CChainState& active_chainstate,
```
same, `pool` not needed if you can get it from `active_chainstate`

glozow commented on src/validation.cpp:
```diff
@@ -337,18 +345,20 @@ static void LimitMempoolSize(CTxMemPool& pool, size_t limit, std::chrono::second
 }
 
-static bool IsCurrentForFeeEstimation() EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+static bool IsCurrentForFeeEstimation(CChainState& chainstate) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
nit: not sure why this is `chainstate` and not `active_chainstate` like the others?

glozow reviewed (commented):
code review ACK https://github.com/bitcoin/bitcoin/pull/20750/commits/2eceb88e52e06e5d58c864b67ff65bc91749eeb1, Cirrus timeout should be fixed with #21117 

Not trying to be annoying but leaving a trail of "don't need mempool if you have chainstate" comments in the case it's made a public member here / in the future.

> I think CheckSequenceLocks() needs both active_chainstate.m_chain.Tip() and active_chainstate.CoinsTip()? I may be missing something here.

Oops @dongcarl you're right 🙈

ryanofsky reviewed (approved):
Code review ACK 2eceb88e52e06e5d58c864b67ff65bc91749eeb1. Changes since last review: dropping removeForReorg & CheckInputsFromMempoolAndCache arguments, adding m_active_chainstate asserts, adding whitespace commits

ariard commented:
Code Review ACK 2eceb88, code changes are adding new asserts in few locations and dropping `CChainState` /`CCoinsViewCache` args when only one was useful. Also dropping passing `nMemPoolHeight` through `removeForReorg`.

dongcarl commented:
Pushed 2eceb88e52e06e5d58c864b67ff65bc91749eeb1 -> 15f0042cb77c0dc4e4f137172d6e871dd2bc55a9
- Addressed https://github.com/bitcoin/bitcoin/pull/20750#discussion_r572993634
- Addressed https://github.com/bitcoin/bitcoin/pull/20750#discussion_r573004038
- Rebased due to conflicts with #21062

glozow commented:
reACK https://github.com/bitcoin/bitcoin/commit/15f0042cb77c0dc4e4f137172d6e871dd2bc55a9

Only changes are renaming to `active_chainstate` in `IsCurrentForFeeEstimation` and using `coins_cache` in `LimitMempoolSize`

ryanofsky reviewed (approved):
Code review ACK 15f0042cb77c0dc4e4f137172d6e871dd2bc55a9. A lot of rebase conflicts and LimitMempoolSize param change since last review.

MarcoFalke commented:
Needs rebase

laanwj commented:
>Needs rebase

It does, git manages to merge but currently the build fails with
```
/…/bitcoin/src/test/util/setup_common.cpp:305:44: error: no matching function for call to 'AcceptToMemoryPool'
        const MempoolAcceptResult result = AcceptToMemoryPool(*m_node.mempool.get(), MakeTransactionRef(mempool_txn), /* bypass_limits */ false);
                                           ^~~~~~~~~~~~~~~~~~
/…/bitcoin/src/validation.h:226:21: note: candidate function not viable: requires at least 4 arguments, but 3 were provided
MempoolAcceptResult AcceptToMemoryPool(CChainState& active_chainstate, CTxMemPool& pool, const CTransactionRef& tx,
                    ^
1 error generated.
```

dongcarl commented:
Pushed 15f0042cb77c0dc4e4f137172d6e871dd2bc55a9 -> e8ae1db864b09a47c736631e6cd3f5ec17929850
- Rebased to fix silent conflict as mentioned in https://github.com/bitcoin/bitcoin/pull/20750#issuecomment-781293235

maflcko reviewed (commented):
ACK e8ae1db864b09a47c736631e6cd3f5ec17929850 📣

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e8ae1db864b09a47c736631e6cd3f5ec17929850 📣
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgMzwwAsEQvRsGjLFTPiwkdMs4DZSXcqw2e+bK/3idvLQRv6sJ7vqrS5W3UIqH4
POgM2H1pz6rbusNlvyXaeNIG/C+RGniPVKf0pBwPrK+Wo/PRpTMMTPyl9cUh3hfB
Xb4HhOCux29lfYNQNFUVq7+O9k2hj9634Q5QCNZkE988rIiGZt6hYc38RIMcSeY7
LU6Pbbc7P7eembZfocL2nVonunPvZij0mLmILc3JWX/jt4C3IqJ6uiQP56ALb08d
gUMuyuQibGz6ptK9S5XeLr2/YyTYt7Z5UNCGZtJyxshFA5MdAxcDhMKQZhs2dW1K
w2jtnuWVWFfHgaU1OvsTjaKmLP8UsCwD4QxJzIgaUgFX//5eWBTN60nyvLVp5Ara
0KndogF/nMxu1KXgVcD7R+LuJiSOJivBX8So9l9wtWwTZdBhb8gybTani25Ef2I4
oB0BSi0GsJlhyvSKBjkvxpzlnf1ofDJ9b6Mid9Z/PgUI50rIO2JixDU5WgXoBvRd
ikxxBzeX
=XgwZ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4c3078abec10b46f834a4805961a34857896683a0701dfcccf98cc987f6a924e  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084c3078abec10b46f834a4805961a34857896683a0701dfcccf98cc987f6a924ef010e7e4d8fe1a661ac433cd9483f0e6beae08fff01005df6fdadcde00895028c9941bd32d8808f10460300dacf008a5df6fdeb5aef80a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010581ce1b0bb5ec6c64c3a8c310713cd2508f10460300dadf00821ba0805e3fc29b90083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01073ccb3d38ce46f0b48504ca75c6e3aed08f120ee3c6f5de9bf15fd8bd1624786b2190705b24bde89ab86b25ab9c27e30d74c2f08f10460300dacf0085a246ae0401038d30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01083e519eb65b50b7f8c67f8b23fa6585908f104602fffdbf00893013508b226fa480083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

glozow commented:
reACK https://github.com/bitcoin/bitcoin/commit/e8ae1db864b09a47c736631e6cd3f5ec17929850 via `git range-diff 15f0042...e8ae1db`, only change is fixing ATMP call from conflict

dongcarl commented on src/validation.cpp:
```diff
@@ -205,8 +205,14 @@ static FlatFileSeq BlockFileSeq();
+    return CheckFinalTx(::ChainActive().Tip(), tx, flags);
+}
+
+bool CheckFinalTx(const CBlockIndex* active_chain_tip, const CTransaction &tx, int flags)
```
Right, this makes sense. I'm going to put this nit off until the next bundle so that I don't invalidate ACKs if that's alright with you!

dongcarl commented on src/txmempool.cpp:
```diff
@@ -356,7 +356,7 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
 {
     // Add to memory pool without checking anything.
-    // Used by AcceptToMemoryPool(), which DOES do
+    // Used by AcceptToMemoryPool(::ChainstateActive(), ), which DOES do
```
Yeah the git syntax was confusing so I wanted to just use `grep` since it's more commonly used 😬
